File: cs/competitive_programming/leetcode_fulltimes/2023/july/33_search_in_rotated_sorted_array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {

        int l=0, r=nums.size()-1;
        int m=0;

        while(l<=r) {
            m = l+(r-l)/2;

            if(nums[m]==target) {
                return m;
            }

            // mid point is on left side
            if(nums[m]>=nums[l]) {

                if( (target>nums[m]) || (target<nums[l]) ) {
                    l = m+1;
                } else {
                    r = m-1;
                }

            }
            else { // mid point is on right side

                if( (target<nums[m]) || (target>nums[r])) {
                    r = m-1;
                } else {
                    l = m+1;
                }


            } 


        }

        return -1;
        
    }
};

File: cs/competitive_programming/leetcode_interns/Blind/Arrays/33_search_in_rotated_sorted_array.py
#!/usr/bin/python3

# https://leetcode.com/problems/search-in-rotated-sorted-array/

from typing import List

class Solution:

    def search(self, nums: List[int], target: int) -> int:
        
        start, end = 0, len(nums)-1

        while start <= end:

            mid = (start+end)//2
            # mid = start + (end-start)//2


            if nums[mid] == target:
                return mid

            # Mid in Left

            if nums[mid] >= nums[start]:

                if target > nums[mid] or target < nums[start]:
                    start = mid+1
                else:
                    end = mid-1
                
            # Mid in Right
            else:
                if target < nums[mid] or target > nums[end]:
                    end = mid-1
                else:
                    start = mid+1

        return -1


obj = Solution()
# nums, target = [4,5,6,7,0,1,2], 0
# out = obj.search(nums, target)
# print(out)

# nums, target = [4,5,6,7,0,1,2], 3
# out = obj.search(nums, target)
# print(out)

nums, target = [1], 0
out = obj.search(nums, target)
print(out)